server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -400,15 +410,38 @@ private void maybeLogHeapDetails() {
      * scans the node paths and loads existing metadata file. If not found a new meta data will be generated
      * and persisted into the nodePaths
      */
-    private static NodeMetadata loadOrCreateNodeMetadata(Settings settings, Logger logger,
-                                                         NodePath... nodePaths) throws IOException {
+    private static NodeMetadata loadNodeMetadata(Settings settings, Logger logger,
+                                                 NodePath... nodePaths) throws IOException {
         final Path[] paths = Arrays.stream(nodePaths).map(np -> np.path).toArray(Path[]::new);
-        NodeMetadata metadata = NodeMetadata.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY, paths);
+        NodeMetadata metadata = PersistedClusterStateService.nodeMetadata(paths);
         if (metadata == null) {
-            metadata = new NodeMetadata(generateNodeId(settings));
+            // load legacy metadata
+            final Set<String> nodeIds = new HashSet<>();
+            for (final Path path : paths) {
+                final NodeMetadata oldStyleMetadata = NodeMetadata.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY, path);
+                if (oldStyleMetadata != null) {
+                    nodeIds.add(oldStyleMetadata.nodeId());
+                }
+            }
+
+            if (nodeIds.size() > 1) {
+                // Technically this can never happen with our our current state because the path will always be prefixed with 'node/$nodeid'
+                // and therefore will be never multiple metadata files within this folder. However, i would still keep this for consistency
+                // reasons
+                throw new IllegalStateException(
+                    "data paths " + Arrays.toString(paths) + " belong to multiple nodes with IDs " + nodeIds);
+            }
+            // load legacy metadata
+            final NodeMetadata legacyMetadata = NodeMetadata.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY, paths);
+            if (legacyMetadata == null) {
+                assert nodeIds.isEmpty() : nodeIds;
+                metadata = new NodeMetadata(generateNodeId(settings));
+            } else {
+                assert nodeIds.equals(Collections.singleton(legacyMetadata.nodeId())) : nodeIds + " doesn't match " + legacyMetadata;
+                metadata = legacyMetadata;
+            }
         }
-        // we write again to make sure all paths have the latest state file
```

Comment:
In the ES version there is a 

```
        metaData = metaData.upgradeToCurrentVersion();
        assert metaData.nodeVersion().equals(Version.CURRENT) : metaData.nodeVersion() + " != " + Version.CURRENT;
```

Don't we need this?

Reply:
Our version of the nodeMetaData does not have a version and upgradeToCurrentVersion yet. Therefore i left it out.